### PR TITLE
Abstract parser registration for compilation

### PR DESCRIPTION
**Pull request by twsouthwick**
This allows the parser to control more of the logic for things such as generating code and identifying dependencies. To add a new "parser" now just requires registering it in the PageCompilationOptions

### CODE REVIEW

**Comment by twsouthwick** [comment]
@birojnayak this should also help with enabling user controls

**Comment by birojnayak** [comment]
taking a look

File: src/Compiler.Dynamic/SystemWebCompilation.cs [hunk 1]
```diff
@@ -165,20 +165,12 @@ private async Task<T[]> GetDependencyPages(TemplateParser parser, CancellationTo
     {
         var p = new List<Task<T>>();
 
-        foreach (var path in GetDependencyPaths(parser))
+        foreach (var path in parser.GetDependencyPaths())
```
Comment (birojnayak) [inline src/Compiler.Dynamic/SystemWebCompilation.cs:168]: will this get the HeaderFile as dependency ?

Comment (twsouthwick) [reply src/Compiler.Dynamic/SystemWebCompilation.cs:168]: what's "HeaderFile"?

Comment (birojnayak) [reply src/Compiler.Dynamic/SystemWebCompilation.cs:168]: UserControl.. sorry sometime mix with other programming languages.

**Review by birojnayak — APPROVED** [review]
looks good to me

File: src/Compiler.Dynamic/SystemWebCompilation.cs (hunk 1 above)
Comment (twsouthwick) [reply src/Compiler.Dynamic/SystemWebCompilation.cs:168]: no - I didn't add any explicit support for usercontrol stuff, but should be a lot easier now

**Pull request merged**